balancer-js/src/modules/pools/pool-types/concerns/composableStable/join.concern.ts

```diff
@@ -109,16 +144,162 @@ export class ComposableStablePoolJoin implements JoinConcern {
+    const assetHelpers = new AssetHelpers(wrappedNativeAsset);
+    // amountsIn must be sorted in correct order for Vault interaction. Currently ordered with relation to tokensIn so need sorted relative to those
+    const [, sortedAmountsIn] = assetHelpers.sortTokens(
+      tokensIn,
+      amountsIn
+    ) as [string[], string[]];
+
+    // This will order everything correctly based on pool tokens
+    const {
+      parsedTokens,
+      parsedAmp,
+      parsedSwapFee,
+      parsedTotalShares,
+      bptIndex,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+    } = parsePoolInfo(pool, wrappedNativeAsset);
+    if (!parsedAmp) {
+      throw new BalancerError(BalancerErrorCode.MISSING_AMP);
+    }
+    return {
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee,
+      bptIndex,
+      parsedTokens,
+    };
+  }
+
+  doJoin(sortedValues: Params): JoinPoolAttributes {
+    const {
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee,
+      bptIndex,
+      parsedTokens,
+      slippage,
+      poolId,
+      joiner,
       value,
-      minBPTOut,
+    } = sortedValues;
+    // BPT out doesn't need downscaled or priceRate
+    const expectedBPTOut = this.doMaths(
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee
+    );
+
+    const userData = this.encodeUserData(
+      expectedBPTOut,
+      slippage,
+      sortedAmountsIn
+    );
+
+    const data = this.encodeFunctionData(
+      poolId,
+      joiner,
+      joiner,
+      parsedTokens,
+      userData.userData,
+      insert(sortedAmountsIn, bptIndex, '0') // Adds value for BPT
+    );
+
+    return {
+      ...data,
+      to: balancerVault,
+      value,
+      minBPTOut: userData.minBPTOut,
     };
-  };
+  }
+
+  // filter native asset (e.g. ETH) amounts
+  getEthValue(tokens: string[], amounts: string[]): BigNumber | undefined {
+    const values = amounts.filter((amount, i) => tokens[i] === AddressZero);
+    return values[0] ? BigNumber.from(values[0]) : undefined;
+  }
+
+  doMaths(
+    amountsIn: string[],
+    scalingFactorsWithoutBpt: bigint[],
+    upScaledBalancesWithoutBpt: string[],
+    upscaledAmp: string,
+    upscaledTotalShares: string,
+    upscaledSwapFee: string
+  ): bigint {
+    /*
+      Maths should use: 
+      - upscaled amounts, e.g. 1USDC = 1e18
+      - rates
+    */
+    const upScaledAmountsIn = _upscaleArray(
+      amountsIn.map(BigInt),
+      scalingFactorsWithoutBpt.map(BigInt)
+    );
+    const expectedBPTOut = StableMathBigInt._calcBptOutGivenExactTokensIn(
+      BigInt(upscaledAmp),
+      upScaledBalancesWithoutBpt.map(BigInt), // Should not have BPT
+      upScaledAmountsIn, // Should not have BPT
```

Comment:
Balances and amounts still need to be multiplied by priceRate, right?

Reply:
Yeah, that's mentioned in comments above. ScalingFactors should include priceRate (same as sc). `parsePoolInfo` has been changed to reflect that.